packages/main/src/TextArea.ts

```diff
@@ -0,0 +1,710 @@
+				"ui5-content-native-scrollbars": getEffectiveScrollbarStyle(),
+			},
+			valueStateMsg: {
+				"ui5-valuestatemessage--error": this.valueState === ValueState.Error,
+				"ui5-valuestatemessage--warning": this.valueState === ValueState.Warning,
+				"ui5-valuestatemessage--information": this.valueState === ValueState.Information,
+			},
+		};
+	}
+
+	get styles() {
+		return {
+			valueStateMsgPopover: {
+				"max-width": `${this._width!}px`,
+			},
+		};
+	}
+
+	get tabIndex() {
+		return this.disabled ? -1 : 0;
+	}
+
+	get ariaLabelText() {
+		const effectiveAriaLabelText = getEffectiveAriaLabelText(this) || getAssociatedLabelForTexts(this);
+
+		if (this.showExceededText) {
+			if (effectiveAriaLabelText) {
+				return effectiveAriaLabelText.concat(" ", this._exceededTextProps.exceededText!);
+			}
+
+			return this._exceededTextProps.exceededText;
+		}
+
+		return effectiveAriaLabelText;
+	}
+
+	get ariaDescribedBy() {
+		return this.hasValueState ? `${this._id}-valueStateDesc` : undefined;
+	}
+
+	get ariaValueStateHiddenText() {
+		if (!this.hasValueState) {
+			return;
+		}
+
+		if (this.hasCustomValueState) {
+			return `${this.valueStateTypeMappings[this.valueState]}`.concat(" ", this.valueStateMessageText.map(el => el.textContent).join(" "));
+		}
+
+		return `${this.valueStateTypeMappings[this.valueState]} ${this.valueStateDefaultText}`;
+	}
+
+	get valueStateDefaultText() {
+		return this.valueStateTextMappings[this.valueState];
+	}
+
+	get ariaInvalid() {
+		return this.valueState === "Error" ? "true" : null;
+	}
+
+	get openValueStateMsgPopover() {
+		return !this._firstRendering && this._openValueStateMsgPopover && this.displayValueStateMessagePopover;
+	}
+
+	get displayValueStateMessagePopover() {
+		return !this.readonly && (this.hasCustomValueState || this.hasValueState);
+	}
+
+	get hasCustomValueState() {
+		return !!this.valueStateMessage.length && this.hasValueState;
+	}
+
+	get hasValueState() {
+		return this.valueState === ValueState.Error || this.valueState === ValueState.Warning || this.valueState === ValueState.Information;
+	}
+
+	get valueStateMessageText() {
+		return this.valueStateMessage.map(x => x.cloneNode(true));
+	}
+
+	get _valueStatePopoverHorizontalAlign() {
+		return this.effectiveDir !== "rtl" ? "Left" : "Right";
+	}
+
+	/**
+	 * This method is relevant for sap_horizon theme only
+	 */
+	get _valueStateMessageIcon() {
+		const iconPerValueState = {
+			Error: "error",
+			Warning: "alert",
+			Success: "sys-enter-2",
+			Information: "information",
+		};
+
+		return this.valueState !== ValueState.None ? iconPerValueState[this.valueState] : "";
+	}
+
+	get valueStateTextMappings() {
+		return {
+			"Success": TextArea.i18nBundle.getText(VALUE_STATE_SUCCESS),
+			"Information": TextArea.i18nBundle.getText(VALUE_STATE_INFORMATION),
+			"Error": TextArea.i18nBundle.getText(VALUE_STATE_ERROR),
+			"Warning": TextArea.i18nBundle.getText(VALUE_STATE_WARNING),
+			"None": "",
```

Comment:
i would prefer to add check where `valueStateTextMappings` is called instead of adding empty scenario for `ValueState.None`